File: src/com/maddyhome/idea/vim/helper/TestInputModel.java
/*
 * IdeaVim - Vim emulator for IDEs based on the IntelliJ platform
 * Copyright (C) 2003-2021 The IdeaVim authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package com.maddyhome.idea.vim.helper;

import com.google.common.collect.Lists;
import com.intellij.openapi.editor.Editor;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.util.List;

/**
 * @author vlan
 */
public class TestInputModel {
  private final @NotNull List<KeyStroke> myKeyStrokes = Lists.newArrayList();

  private TestInputModel() {}

  public static TestInputModel getInstance(@NotNull Editor editor) {
    TestInputModel model = UserDataManager.getVimTestInputModel(editor);
    if (model == null) {
      model = new TestInputModel();
      UserDataManager.setVimTestInputModel(editor, model);
    }
    return model;
  }

  public void setKeyStrokes(@NotNull List<KeyStroke> keyStrokes) {
    myKeyStrokes.clear();
    myKeyStrokes.addAll(keyStrokes);
  }

  public @Nullable KeyStroke nextKeyStroke() {
    if (!myKeyStrokes.isEmpty()) {
      return myKeyStrokes.remove(0);
    }
    return null;
  }
}

File: src/main/java/com/maddyhome/idea/vim/helper/TestInputModel.kt
/*
 * Copyright 2003-2023 The IdeaVim authors
 *
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE.txt file or at
 * https://opensource.org/licenses/MIT.
 */
package com.maddyhome.idea.vim.helper

import com.google.common.collect.Lists
import com.intellij.openapi.editor.Editor
import com.maddyhome.idea.vim.key.VimKeyStroke

// Do not remove until it's used in EasyMotion plugin in tests
class TestInputModel private constructor() {
  private val myKeyStrokes: MutableList<VimKeyStroke> = Lists.newArrayList()
  fun setKeyStrokes(keyStrokes: List<VimKeyStroke>) {
    myKeyStrokes.clear()
    myKeyStrokes.addAll(keyStrokes)
  }

  fun nextKeyStroke(): VimKeyStroke? {
    // Return key from the unfinished mapping
    /*
MappingStack mappingStack = KeyHandler.getInstance().getMappingStack();
if (mappingStack.hasStroke()) {
  return mappingStack.feedStroke();
}
*/
    return if (myKeyStrokes.isNotEmpty()) {
      myKeyStrokes.removeAt(0)
    } else {
      null
    }
  }

  companion object {
    @JvmStatic
    fun getInstance(editor: Editor): TestInputModel {
      var model = editor.vimTestInputModel
      if (model == null) {
        model = TestInputModel()
        editor.vimTestInputModel = model
      }
      return model
    }
  }
}
